crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,566 @@
+use crate::virtio_sound::{
+    VirtioSndPcmSetParams, VIRTIO_SND_D_INPUT, VIRTIO_SND_D_OUTPUT, VIRTIO_SND_PCM_FMT_A_LAW,
+    VIRTIO_SND_PCM_FMT_FLOAT, VIRTIO_SND_PCM_FMT_FLOAT64, VIRTIO_SND_PCM_FMT_MU_LAW,
+    VIRTIO_SND_PCM_FMT_S16, VIRTIO_SND_PCM_FMT_S18_3, VIRTIO_SND_PCM_FMT_S20,
+    VIRTIO_SND_PCM_FMT_S20_3, VIRTIO_SND_PCM_FMT_S24, VIRTIO_SND_PCM_FMT_S24_3,
+    VIRTIO_SND_PCM_FMT_S32, VIRTIO_SND_PCM_FMT_S8, VIRTIO_SND_PCM_FMT_U16,
+    VIRTIO_SND_PCM_FMT_U18_3, VIRTIO_SND_PCM_FMT_U20, VIRTIO_SND_PCM_FMT_U20_3,
+    VIRTIO_SND_PCM_FMT_U24, VIRTIO_SND_PCM_FMT_U24_3, VIRTIO_SND_PCM_FMT_U32,
+    VIRTIO_SND_PCM_FMT_U8, VIRTIO_SND_PCM_RATE_11025, VIRTIO_SND_PCM_RATE_16000,
+    VIRTIO_SND_PCM_RATE_176400, VIRTIO_SND_PCM_RATE_192000, VIRTIO_SND_PCM_RATE_22050,
+    VIRTIO_SND_PCM_RATE_32000, VIRTIO_SND_PCM_RATE_384000, VIRTIO_SND_PCM_RATE_44100,
+    VIRTIO_SND_PCM_RATE_48000, VIRTIO_SND_PCM_RATE_5512, VIRTIO_SND_PCM_RATE_64000,
+    VIRTIO_SND_PCM_RATE_8000, VIRTIO_SND_PCM_RATE_88200, VIRTIO_SND_PCM_RATE_96000,
+    VIRTIO_SND_S_BAD_MSG, VIRTIO_SND_S_NOT_SUPP,
+};
+use crate::Error;
+//equiv of PCMParams
+use crate::Stream;
+use std::{
+    cmp,
+    collections::HashMap,
+    convert::TryInto,
+    ops::Deref,
+    ptr,
+    ptr::NonNull,
+    sync::{Arc, RwLock},
+};
 
+use log::debug;
+use pipewire as pw;
+use pw::spa::param::ParamType;
+use pw::spa::pod::Pod;
+use std::ffi::CString;
+use std::mem::size_of;
+
+use pw::spa::sys::{
+    spa_audio_info_raw, SPA_AUDIO_CHANNEL_FC, SPA_AUDIO_CHANNEL_FL, SPA_AUDIO_CHANNEL_FR,
+    SPA_AUDIO_CHANNEL_LFE, SPA_AUDIO_CHANNEL_MONO, SPA_AUDIO_CHANNEL_RC, SPA_AUDIO_CHANNEL_RL,
+    SPA_AUDIO_CHANNEL_RR, SPA_AUDIO_CHANNEL_UNKNOWN, SPA_AUDIO_FORMAT_ALAW, SPA_AUDIO_FORMAT_F32,
+    SPA_AUDIO_FORMAT_F64, SPA_AUDIO_FORMAT_S16, SPA_AUDIO_FORMAT_S18_LE, SPA_AUDIO_FORMAT_S20,
+    SPA_AUDIO_FORMAT_S20_LE, SPA_AUDIO_FORMAT_S24, SPA_AUDIO_FORMAT_S24_LE, SPA_AUDIO_FORMAT_S32,
+    SPA_AUDIO_FORMAT_S8, SPA_AUDIO_FORMAT_U16, SPA_AUDIO_FORMAT_U18_LE, SPA_AUDIO_FORMAT_U20,
+    SPA_AUDIO_FORMAT_U20_LE, SPA_AUDIO_FORMAT_U24, SPA_AUDIO_FORMAT_U24_LE, SPA_AUDIO_FORMAT_U32,
+    SPA_AUDIO_FORMAT_U8, SPA_AUDIO_FORMAT_ULAW, SPA_AUDIO_FORMAT_UNKNOWN,
+};
+
+use pw::sys::{
+    pw_buffer, pw_loop, pw_thread_loop, pw_thread_loop_get_loop, pw_thread_loop_lock,
+    pw_thread_loop_new, pw_thread_loop_signal, pw_thread_loop_start, pw_thread_loop_unlock,
+    pw_thread_loop_wait, PW_ID_CORE,
+};
+use pw::{properties, spa, Context, Core, LoopRef};
+
+#[macro_export]
+macro_rules! properties_set {
+    ($properties:expr, $key:expr, $($args:tt)*) => {{
+        let key_cstr = CString::new($key).unwrap();
+        let format_str = format!($($args)*);
+        let format_cstr = CString::new(format_str).unwrap();
+        unsafe {
+            pw::sys::pw_properties_set(
+                $properties.as_ptr(),
+                key_cstr.as_ptr(),
+                format_cstr.as_ptr(),
+            )
+        }
+    }};
```

Comment:
Any reason of using a macro here instead of function?  maybe this should be also upstreamed and be part of `pipewire::Properties`

Reply:
I wanted to upstream this macro but the pipewire-rs upstream maintainer would not accept it, and suggested instead for users to set the format macro themselves in their projects.

Reply:
This should really be a function imho.

Reply:
but even the C function that you are wrapping here is not using a variadic argument, doing so in rust just hurts readability and macros are terrible for LSPs (language servers providers like rust analyzer)